ref/haskell/src/Lib.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Lib
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8)
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+charsetMap :: Map.HashMap Char Word8
+charsetMap = Map.fromList $ zip charset [0..]
+
+charsetMapInv :: Map.HashMap Word8 Char
+charsetMapInv = Map.fromList $ zip [0..] charset
+
+bech32Polymod :: Data -> Int
+bech32Polymod = foldl' go 1
+  where
+    go chk value = foldl' xor chk' [(- (top .>>. i) .&. 1) .&. g | (g, i) <- zip generator [0..]]
+      where
+        generator = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3]
+        top = chk .>>. 25
+        chk' = (chk .&. 0x1ffffff) .<<. 5 `xor` (fromIntegral value)
```

Comment:
Sorry I was a bit hasty.

The transformation is 
```
(chk .&. 0x1ffffff) .<<. 5
```
 = { distribution of .<<. and .&. }
```
(chk .<<. 5) .&. (0x1ffffff .<<. 5)
```
 = { compute (0x1ffffff .<<. 5) }
```
(chk .<<. 5) .&. (0x3fffffe0)
```
= { low bits are 0 anyways }
```
(chk .<<. 5) .&. (0x3fffffff)
```

And now you can float the `0x3fffffff` to the top.

```
bech32Polymod = foldl' go 1 .&. 0x3fffffff
```

I think you are right that the `Word32`ness doesn't matter.  I just didn't want to worry about sign extensions when thinking about it.